python/cudf/cudf/_lib/ast.pyx

```diff
@@ -0,0 +1,348 @@
+                op_value, dereference(left.c_obj)
+            )
+        else:
+            self.c_obj = <expression_ptr> make_unique[libcudf_ast.operation](
+                op_value, dereference(left.c_obj), dereference(right.c_obj)
+            )
+
+
+# This dictionary encodes the mapping from Python AST operators to their cudf
+# counterparts.
+python_cudf_operator_map = {
+    # Binary operators
+    ast.Add: ASTOperator.ADD,
+    ast.Sub: ASTOperator.SUB,
+    ast.Mult: ASTOperator.MUL,
+    ast.Div: ASTOperator.DIV,
+    ast.FloorDiv: ASTOperator.FLOOR_DIV,
+    ast.Mod: ASTOperator.PYMOD,
+    ast.Pow: ASTOperator.POW,
+    ast.Eq: ASTOperator.EQUAL,
+    ast.NotEq: ASTOperator.NOT_EQUAL,
+    ast.Lt: ASTOperator.LESS,
+    ast.Gt: ASTOperator.GREATER,
+    ast.LtE: ASTOperator.LESS_EQUAL,
+    ast.GtE: ASTOperator.GREATER_EQUAL,
+    ast.BitXor: ASTOperator.BITWISE_XOR,
+    # TODO: The mapping of logical/bitwise operators here is inconsistent with
+    # pandas. In pandas, Both `BitAnd` and `And` map to
+    # `ASTOperator.LOGICAL_AND` for booleans, while they map to
+    # `ASTOperator.BITWISE_AND` for integers. However, there is no good way to
+    # encode this at present because expressions can be arbitrarily nested so
+    # we won't know the dtype of the input without inserting a much more
+    # complex traversal of the expression tree to determine the output types at
+    # each node. For now, we'll rely on users to use the appropriate operator.
+    ast.BitAnd: ASTOperator.BITWISE_AND,
+    ast.BitOr: ASTOperator.BITWISE_OR,
+    ast.And: ASTOperator.LOGICAL_AND,
+    ast.Or: ASTOperator.LOGICAL_OR,
+
+    # Unary operators
+    ast.Invert: ASTOperator.BIT_INVERT,
+    ast.Not: ASTOperator.NOT,
+    # TODO: Missing USub, possibility other unary ops?
+}
+
+
+# Mapping between Python function names encode in an ast.Call node and the
+# corresponding libcudf C++ AST operators.
+python_cudf_function_map = {
+    # TODO: Operators listed on
+    # https://pandas.pydata.org/pandas-docs/stable/user_guide/enhancingperf.html#expression-evaluation-via-eval
+    # that we don't support yet:
+    # expm1, log1p, arctan2 and log10.
+    'sin': ASTOperator.SIN,
+    'cos': ASTOperator.COS,
+    'tan': ASTOperator.TAN,
+    'arcsin': ASTOperator.ARCSIN,
+    'arccos': ASTOperator.ARCCOS,
+    'arctan': ASTOperator.ARCTAN,
+    'sinh': ASTOperator.SINH,
+    'cosh': ASTOperator.COSH,
+    'tanh': ASTOperator.TANH,
+    'arcsinh': ASTOperator.ARCSINH,
+    'arccosh': ASTOperator.ARCCOSH,
+    'arctanh': ASTOperator.ARCTANH,
+    'exp': ASTOperator.EXP,
+    'log': ASTOperator.LOG,
+    'sqrt': ASTOperator.SQRT,
+    'abs': ASTOperator.ABS,
+
+    # TODO: Operators supported by libcudf with no Python function analog.
+    # ast.rint: ASTOperator.RINT,
+    # ast.cbrt: ASTOperator.CBRT,
+    # ast.ceil: ASTOperator.CEIL,
+    # ast.floor: ASTOperator.FLOOR,
+}
+
+
+class libcudfASTVisitor(ast.NodeVisitor):
```

Comment:
I would consider moving this entire class to a `.py` file. It would make it much easier to debug and potentially extend later. 

Locally I managed to migrate this class and `parse_expression` to a utils and then got it to pass tests importing from there into `transform.pyx`. After chatting offline with @shwina we thought maybe it would be reasonable to put that `.py` file directly inside `cudf._lib`, since it would keep the cython API self contained and avoid importing into the cython layer from the cuDF python layer.

Reply:
There's no reason `cudf._lib` cannot contain `.py` components. I agree that larger, pure Python classes like this would benefit from living in pure Python files for ease of debugging.

Reply:
I am also definitely in favor of this and will make the change.